jina/peapods/grpc/__init__.py

```diff
@@ -0,0 +1,148 @@
+import inspect
+from typing import Optional, Callable
+
+import grpc
+from google.protobuf import struct_pb2
+
+from jina.proto import jina_pb2_grpc, jina_pb2
+from jina.types.message import Message
+from jina.types.message.common import ControlMessage
+from jina.types.routing.table import RoutingTable
+
+
+# TODO extract dynamic routing logic and merge with zmqlet
+class Grpclet(jina_pb2_grpc.JinaDataRequestRPCServicer):
+    """A `Grpclet` object can send/receive Messages via gRPC.
+
+    :param async_loop: asyncio loop to use
+    :param args: the parsed arguments from the CLI
+    :param message_callback: the callback to call on received messages
+    :param logger: the logger to use
+    """
+
+    def __init__(
+        self,
+        async_loop,
+        args: 'argparse.Namespace',
+        message_callback: Callable[['Message'], 'Message'],
+        logger: Optional['JinaLogger'] = None,
+    ):
+        self.args = args
+        self._stubs = {}
+        self._logger = logger
+        self.callback = message_callback
+        self._loop = async_loop
+        self.msg_recv = 0
+        self.msg_sent = 0
+
+    async def send_message(self, msg: 'Message', **kwargs):
+        """
+        Sends a message via gRPC to the target indicated in the message's routing table
+        :param msg: the protobuf message to send
+        :param kwargs: Additional arguments.
+        """
+        routing_table = RoutingTable(msg.envelope.routing_table)
+        next_targets = routing_table.get_next_targets()
+
+        for target, _ in next_targets:
+            pod_address = target.active_target_pod.full_address
+
+            new_message = await self._add_envelope(msg, target)
+
+            if pod_address not in self._stubs:
+                self._stubs[pod_address] = Grpclet._create_grpc_stub(pod_address)
+
+            try:
+                self.msg_sent += 1
+                self._stubs[pod_address].Call(new_message)
```

Comment:
`_create_grpc_stub` by default seems to be async, hence this should be awaited?

Reply:
Not awaited (if we dont want to wait actually), but I had to wrap it actually in a task